volcli/volcli.go

```diff
@@ -286,6 +289,123 @@ func policyList(ctx *cli.Context) (bool, error) {
 	return false, nil
 }
 
+// PolicyGetRevision retrieves a single revision from a policy's history.
+func PolicyGetRevision(ctx *cli.Context) {
+	execCliAndExit(ctx, policyGetRevision)
+}
+
+func policyGetRevision(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 2 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 2, ctx.Args())
+	}
+
+	name := ctx.Args()[0]
+	revision := ctx.Args()[1]
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/policy-archives/%s/%s",
+		ctx.GlobalString("apiserver"),
+		name,
+		revision,
+	))
+	if err != nil {
+		return false, err
+	}
+
+	if resp.StatusCode != 200 {
+		if _, err := io.Copy(os.Stderr, resp.Body); err != nil {
+			return false, errored.Errorf("Error copying body: %v\nResponse Status Code was %d, not 200", err, resp.StatusCode)
+		}
+		return false, errored.Errorf("Response Status Code was %d, not 200", resp.StatusCode)
+	}
+
+	content, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+
+	fmt.Println(string(content))
+
+	return false, nil
+}
+
+// PolicyListRevisions retrieves all the revisions for a given policy.
+func PolicyListRevisions(ctx *cli.Context) {
+	execCliAndExit(ctx, policyListRevisions)
+}
+
+func policyListRevisions(ctx *cli.Context) (bool, error) {
+	var revisions []string
+
+	if len(ctx.Args()) != 1 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 1, ctx.Args())
+	}
+
+	name := ctx.Args()[0]
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/policy-archives/%s",
+		ctx.GlobalString("apiserver"),
+		name,
+	))
+	if err != nil {
+		return false, err
+	}
+
+	if resp.StatusCode != 200 {
+		if _, err := io.Copy(os.Stderr, resp.Body); err != nil {
+			return false, errored.Errorf("Error copying body: %v\nResponse Status Code was %d, not 200", err, resp.StatusCode)
+		}
+		return false, errored.Errorf("Response Status Code was %d, not 200", resp.StatusCode)
+	}
+
+	content, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+
+	if err := json.Unmarshal(content, &revisions); err != nil {
+		return false, err
+	}
+
+	for _, revision := range revisions {
+		fmt.Println(revision)
+	}
+
+	return false, nil
+}
+
+// PolicyWatch watches etcd for policy changes and prints them.
+func PolicyWatch(ctx *cli.Context) {
+	execCliAndExit(ctx, policyWatch)
+}
+
+func policyWatch(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 0 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 0, ctx.Args())
+	}
+
+	cfg, err := config.NewClient(ctx.GlobalString("prefix"), ctx.GlobalStringSlice("etcd"))
+	if err != nil {
+		return false, err
+	}
+
+	wc := make(chan *watch.Watch)
+
+	cfg.WatchForPolicyChanges(wc, func(resp *client.Response, w *watch.Watcher) {
+		w.Channel <- &watch.Watch{Key: resp.Node.Key, Config: true}
+	})
+
+	for w := range wc {
+		keyspace, revision := path.Split(w.Key)
+		keyspace = keyspace[:len(keyspace)-1] // remove trailing slash
+
+		_, name := path.Split(keyspace)
```

Comment:
ok, this one in particular confuses me.

If the path is `/volplugin/policy-archives/policy1/test/<timestamp>`, this will yield the name as `test` not `policy1/test` if I'm reading this right. Is that accurate? If so, we need to change it to `policy1/test`.

If that's how this works disregard, I just don't have bandwidth to test the patch myself.